lib/bundler/plugin.rb

```diff
@@ -121,13 +123,35 @@ def source_from_lock(locked_opts)
       src.new(locked_opts.merge("uri" => locked_opts["remote"]))
     end
 
+    def add_hook(event, &block)
+      (@hooks[event.to_s] ||= []) << block
+    end
+
+    def hook(event, *args)
+      plugins = index.hook_plugins(event)
+      return unless plugins
+
+      (plugins - @loaded).each {|name| load_plugin(name) }
+
+      @hooks[event].each {|blk| blk.call(*args) }
+    end
+
     # currently only intended for specs
     #
     # @return [String, nil] installed path
     def installed?(plugin)
       Index.new.installed?(plugin)
     end
 
+    def reset!
```

Comment:
maybe just call `reset!` instead of having the separate ivar initialization above?